File: trash/monte_carlo.py
from board_wrapper import Move, Board
import math

class Node():

    def __init__(self):
        self.total = 0
        self.visits = 0
        self.legal_moves = None
        self.logits = None
        self.evaluation = None
        self.game_ended = False


class MonteCarlo():

    def __init__(self, model, iterations=50000, max_depth=250, max_nodes=500000):
        self.nodes = {}
        self.model = model
        self.nodes_visited = 0
        self.iterations = iterations
        self.max_depth=250

        self.node_pointer = 0
        self.max_nodes = max_nodes
        self.hash_list = [0]*max_nodes

    def search(self, board: Board):

        self.nodes_visited = 0

        for i in range(self.iterations):
            self.visit(board)

        # Print total iterations per second at the end

        node = self.get_node(board)

        if board.get_winner() != 0 or len(node.legal_moves) == 0:
            return None

        best = float("-inf")
        move = None

        for m in board.get_legal_moves():

            board.play(m)
            child_node = self.get_node(board)
            board.undo(m)

            #print(m, child_node.visits)

            score = child_node.visits
            if score > best:
                move = m
                best = score
        
        return move

    def get_node(self, board):
        h = board.get_hash()
        if not (h in self.nodes):
            
            if self.hash_list[self.node_pointer] in self.nodes:
                self.nodes.pop(self.hash_list[self.node_pointer])

            self.nodes[h] = Node()
            self.hash_list[self.node_pointer] = h
            self.node_pointer = (self.node_pointer + 1) % self.max_nodes

        return self.nodes[h]

    def roll_out(self, node: Node, board: Board):
        if node.legal_moves == None:
            node.legal_moves = board.get_legal_moves()

        node.game_ended = (
            board.get_winner() != 0
            or len(node.legal_moves) == 0
            or board.is_repetition()
            or board.is_insufficient())

        if node.game_ended:
            winner = board.get_winner()

            if winner != 0:
                if winner == -1:
                    node.evaluation = -1
                else:
                    node.evaluation = 1
            else:
                node.evaluation = 0
        else:
            if node.logits == None or node.evaluation == None:
                evaluation, logits = self.model(board, node.legal_moves)
                node.logits = logits
                node.evaluation = evaluation

        

    
    def node_weight(self, node, N, white_turn, depth): #uses modified UCB1
        if node.visits == 0:
            return float("inf")

        weight = node.total/node.visits
        adj =  math.exp(-depth/5) * math.sqrt(math.log(N*2)/node.visits)
        if white_turn:
            return -weight + adj
        return weight + adj


    def visit(self, board: Board, depth=1):
        
        self.nodes_visited += 1

        if depth >= self.max_depth:
            return 0 #if game carries on for too long, it is probably a tie

        node = self.get_node(board)

        if node.visits == 0: # leaf node
            self.roll_out(node, board)
            node.visits += 1
            node.total += node.evaluation
            return node.evaluation
        
        if node.game_ended:
            node.visits += 1
            node.total += node.evaluation
            return node.evaluation
        
        best_weight = float("-inf")
        move = None
        
        i = 0

        legals = node.legal_moves
        for m in legals:
            
            board.play(m)
            weight = self.node_weight(self.get_node(board), node.visits, board.is_white_turn(), depth) 
            board.undo(m)

            if weight > best_weight:
                move = m
                best_weight = weight
            i += 1

        board.play(move)
        evaluation = self.visit(board, depth+1) #back propagation
        board.undo(move)

        node.total += evaluation
        node.visits += 1

        return evaluation
    

File: src/monte_carlo.cpp
#include "monte_carlo.h"
#include "timer.h"
#include "model.h"
#include <cmath>
#include <algorithm>
#include <random>
#include <numeric>


template<typename T>
inline int random_index(const std::vector<T>& weights) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0.0, 1.0);

    float total = std::accumulate(weights.begin(), weights.end(), 0.0f);
    float random_value = dis(gen);

    float cumulative_sum = 0.0f;
    for (size_t i = 0; i < weights.size(); ++i) {
        cumulative_sum += weights[i] / total;
        if (random_value < cumulative_sum) {
            return i;
        }
    }
    return weights.size() - 1;
}

template<typename T>
inline int max_index(const std::vector<T>& vec) {
    if (vec.size() == 0) {
        throw std::invalid_argument("Input vector to max_index is empty.");
    }

    auto max_it = std::max_element(vec.begin(), vec.end());
    return std::distance(vec.begin(), max_it);
}

void calculateZScores(std::vector<float>& data) {
    // Step 1: Calculate the mean
    float sum = std::accumulate(data.begin(), data.end(), 0.0f);
    float mean = sum / data.size();

    // Step 2: Calculate the standard deviation
    float sq_sum = std::accumulate(data.begin(), data.end(), 0.0f, 
                                   [mean](float acc, float x) { return acc + (x - mean) * (x - mean); });
    float std_dev = std::sqrt(sq_sum / data.size());

    // Step 3: Transform each element into its Z-score
    for (float& x : data) {
        x = (x - mean) / std_dev;
    }
}



bool is_white_king_dead(Board& board) { return board.piece_bitboard(WHITE_KING) == 0; }
bool is_black_king_dead(Board& board) { return board.piece_bitboard(BLACK_KING) == 0; }
bool is_game_draw(Board& board) { return board.is_insufficient() || board.is_repetition() || board.is_rule_50(); }





MonteCarlo::MonteCarlo(Model& m, MonteCarloConfig config) : model(m), nodes_map{} {
    this->is_black_win = is_white_king_dead;
    this->is_white_win = is_black_king_dead;
    this->is_draw = is_game_draw;
    this->exploration_scale = config.exploration_scale;
    this->exploration_decay = config.exploration_decay;
    this->max_nodes = config.max_nodes;
    this->max_depth = config.max_depth;
}


MonteCarlo::MonteCarlo(Model& m) : MonteCarlo(m, MonteCarloConfig()) {}

inline Node& MonteCarlo::get_node(Board& board) {

    uint64_t hash = board.get_hash();

    if (this->nodes_map.find(hash) != this->nodes_map.end()) {
        return this->nodes_map[hash];
    }
    
    this->nodes_map[hash] = Node(); 
    return this->nodes_map[hash];
}


inline void MonteCarlo::roll_out(Board& board, Node& node) {

    node.legal_moves = board.get_legal_moves();
    node.move_weights = std::vector<float>(node.legal_moves.size(), 1);

    if (this->is_white_win(board)) {
        node.evaluation = 1;
        node.game_ended = true;
    } else if (this->is_black_win(board)) {
        node.evaluation = -1;
        node.game_ended = true;
    } else if (this->is_draw(board)) {
        node.evaluation = 0;
        node.game_ended = true;
    } else {
        
        node.evaluation = this->model(board, node.legal_moves, node.move_weights);
        std::transform(
            node.move_weights.begin(),
            node.move_weights.end(),
            node.move_weights.begin(),
            static_cast<float(*)(float)>(std::exp) 
        );
        //model returns logits, which can be negative so we can take e^x for positive values
        //we are essentially trying to compute softmax later on
    }
}



inline float MonteCarlo::node_weight(Node& node, int N, bool white_turn, int depth) { //uses modified UCB1
    if (node.visits == 0) {
        return INFINITY;
    }

    float exploitation = node.total/float(node.visits);
    float exploration = this->exploration_scale 
                        * std::exp(-depth * this->exploration_decay) 
                        * std::sqrt(std::log(N*2)/float(node.visits)); 
                        //for higher depth, we want to do less exploring and more exploiting
    
    if (white_turn) {
        return -exploitation + exploration;
    }
    return exploitation + exploration;
}



inline float MonteCarlo::visit(Board& board, int depth) {

    if (depth >= this->max_depth) {
        return 0;
    }
        

    Node& node = this->get_node(board);

    if (node.visits == 0) { //leaf node
        this->roll_out(board, node);
        node.visits++;
        node.total += node.evaluation;
        return node.evaluation;
    }

    
    if (node.game_ended | node.legal_moves.size() == 0) {
        node.visits++;
        node.total += node.evaluation;
        return node.evaluation;
    }

    std::vector<Move> unexplored_moves;
    std::vector<float> unexplored_weights;

    std::vector<Move> explored_moves;
    std::vector<float> explored_weights;

    //need to store all the variables before hand, since get_node can delete node pointers
    std::vector<Move> legal_moves = node.legal_moves;
    std::vector<float> move_weights = node.move_weights;
    uint32_t total_visits = node.visits;

    for (int i = 0; i < legal_moves.size(); i++) {
        Move move = legal_moves[i];
        board.play(move);
        Node& n = this->get_node(board);
        if (n.visits <= 0) {
            unexplored_moves.push_back(move);
            unexplored_weights.push_back(move_weights[i]);
        } else {
            float weight = this->node_weight(n, total_visits, board.turn()==WHITE, depth);
            explored_moves.push_back(move);
            explored_weights.push_back(weight);
        }

        board.undo(move);
    }

    Move best_move;

    if (unexplored_moves.size() > 0) {
        best_move = unexplored_moves[random_index<float>(unexplored_weights)];
    } else {
        best_move = explored_moves[max_index<float>(explored_weights)];
    }


    board.play(best_move);
    float eval = this->visit(board, depth+1); //back propagation
    board.undo(best_move);
    
    node = this->get_node(board);
    node.total += eval;
    node.visits++;

    return eval;
}


Move MonteCarlo::search(Board& board, int search_time_ms) {
    
    if (board.get_legal_moves().size() == 0) {
        throw std::invalid_argument("MonteCarlo.search() can not be called for positions with no legal moves");
    }
    
    this->iterations_searched = 0;

    if (this->is_draw(board) || this->is_black_win(board) || this->is_white_win(board)) {
        return Move();
    }


    Timer timer(search_time_ms);

    this->nodes_map = std::unordered_map<uint64_t, Node>();

    while (timer.time_remaining() > 0) {
        for (int i = 0; i < 100; i++) {
            this->visit(board, 0);
            this->iterations_searched++;
            if (this->iterations_searched >= this->max_nodes) break;
        }
        if (this->iterations_searched >= this->max_nodes) break;
    }
    
    std::vector<Move> legal_moves = board.get_legal_moves();
    std::vector<float> visits;

    for (Move m : legal_moves) {
        board.play(m);
        Node& child_node = this->get_node(board);
        board.undo(m);
        visits.push_back(child_node.visits);
    }

    //calculateZScores(visits);
    //
    //for (int i = 0; i < visits.size(); i++) {
    //    visits[i] = std::exp(std::min(visits[i] * 5.0f, 2.0f));
    //}

    return legal_moves[max_index<float>(visits)];
}  



int MonteCarlo::get_iterations_searched() {
    return this->iterations_searched;
}